ocaml/parsing/pprintast.ml

```diff
@@ -355,10 +365,16 @@ let maybe_modes_type pty ctxt f c =
   | None -> pty ctxt f c
 
 let modalities_type pty ctxt f pca =
-  match pca.pca_modalities with
-  | [] -> pty ctxt f pca.pca_type
-  | m ->
-    pp f "%a %a" legacy_modalities m (pty ctxt) pca.pca_type
+  let modalities = Split_modalities.split_list pca.pca_modalities in
+  let space_if_legacies =
+    match modalities.legacy with
+    | [] -> ""
+    | _ :: _ -> " "
+  in
```

Comment:
there's already a function `optional_legacy_modalities` that should do the space for you. everything else looks good though, so feel free to merge after fixing